File: src/main/java/com/gazman/db/DbProxy.java
package com.gazman.db;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteTransactionListener;
import android.os.Build;
import android.os.CancellationSignal;
import android.util.Pair;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.sqlite.db.SupportSQLiteQuery;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import io.requery.android.database.sqlite.SQLiteDatabase;
import io.requery.android.database.sqlite.SQLiteStatement;

public class DbProxy implements SupportSQLiteDatabase {
    public static final int CONFLICT_IGNORE = SQLiteDatabase.CONFLICT_IGNORE;
    public static final int CONFLICT_REPLACE = SQLiteDatabase.CONFLICT_REPLACE;
    public static @Nullable
    QueryLogger queryLogger;
    private final SQLiteDatabase db;

    public DbProxy(SQLiteDatabase db) {
        this.db = db;
    }

    @Override
    public SQLiteStatement compileStatement(String sql) {
        return db.compileStatement(sql);
    }

    @Override
    public void beginTransaction() {
        db.beginTransaction();
    }

    @Override
    public void beginTransactionNonExclusive() {
        db.beginTransactionNonExclusive();
    }

    @Override
    public void beginTransactionWithListener(SQLiteTransactionListener transactionListener) {
        db.beginTransactionWithListener(transactionListener);
    }

    @Override
    public void beginTransactionWithListenerNonExclusive(SQLiteTransactionListener transactionListener) {
        db.beginTransactionWithListenerNonExclusive(transactionListener);
    }

    @Override
    public void endTransaction() {
        db.endTransaction();
    }

    @Override
    public void setTransactionSuccessful() {
        db.setTransactionSuccessful();
    }

    @Override
    public boolean inTransaction() {
        return db.inTransaction();
    }

    @Override
    public boolean isDbLockedByCurrentThread() {
        return db.isDbLockedByCurrentThread();
    }

    @Override
    public boolean yieldIfContendedSafely() {
        return db.yieldIfContendedSafely();
    }

    @Override
    public boolean yieldIfContendedSafely(long sleepAfterYieldDelay) {
        return db.yieldIfContendedSafely(sleepAfterYieldDelay);
    }

    @Override
    public int getVersion() {
        return db.getVersion();
    }

    @Override
    public void setVersion(int version) {
        db.setVersion(version);
    }

    @Override
    public long getMaximumSize() {
        return db.getMaximumSize();
    }

    @Override
    public long setMaximumSize(long numBytes) {
        return db.setMaximumSize(numBytes);
    }

    @Override
    public long getPageSize() {
        return db.getPageSize();
    }

    @Override
    public void setPageSize(long numBytes) {
        db.setPageSize(numBytes);
    }

    @Override
    public Cursor query(String query) {
        log(query);
        return db.query(query);
    }

    @Override
    public Cursor query(String query, Object[] bindArgs) {
        log(query);
        return db.query(query, bindArgs);
    }

    @Override
    public Cursor query(SupportSQLiteQuery query) {
        return db.query(query);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    @Override
    public Cursor query(SupportSQLiteQuery query, CancellationSignal cancellationSignal) {
        return db.query(query, cancellationSignal);
    }

    @Override
    public long insert(String table, int conflictAlgorithm, ContentValues values) throws SQLException {
        return db.insert(table, conflictAlgorithm, values);
    }

    public long insert(String table, String nullColumnHack, ContentValues values) {
        return db.insert(table, nullColumnHack, values);
    }

    @Override
    public int delete(String table, String whereClause, Object[] whereArgs) {
        return db.delete(table, whereClause, whereArgs);
    }

    @Override
    public int update(String table, int conflictAlgorithm, ContentValues values, String whereClause, Object[] whereArgs) {
        return db.update(table, conflictAlgorithm, values, whereClause, whereArgs);
    }

    public int update(String table, ContentValues values, String whereClause, String[] whereArgs) {
        return db.update(table, values, whereClause, whereArgs);
    }

    @Override
    public void execSQL(String query) throws SQLException {
        log(query);
        db.execSQL(query);
    }

    @Override
    public void execSQL(String query, Object[] bindArgs) throws SQLException {
        log(query);
        db.execSQL(query);
    }

    private void log(String query) {
        QueryLogger queryLogger = DbProxy.queryLogger;
        if (queryLogger != null) {
            queryLogger.onQuery(query);
        }
    }

    @Override
    public boolean isReadOnly() {
        return db.isReadOnly();
    }

    @Override
    public boolean isOpen() {
        return db.isOpen();
    }

    @Override
    public boolean needUpgrade(int newVersion) {
        return db.needUpgrade(newVersion);
    }

    @Override
    public String getPath() {
        return db.getPath();
    }

    @Override
    public void setLocale(Locale locale) {
        db.setLocale(locale);
    }

    @Override
    public void setMaxSqlCacheSize(int cacheSize) {
        db.setMaxSqlCacheSize(cacheSize);
    }

    @Override
    public void setForeignKeyConstraintsEnabled(boolean enable) {
        db.setForeignKeyConstraintsEnabled(enable);
    }

    @Override
    public boolean enableWriteAheadLogging() {
        return db.enableWriteAheadLogging();
    }

    @Override
    public void disableWriteAheadLogging() {
        db.disableWriteAheadLogging();
    }

    @Override
    public boolean isWriteAheadLoggingEnabled() {
        return db.isWriteAheadLoggingEnabled();
    }

    @Override
    public List<Pair<String, String>> getAttachedDbs() {
        return db.getAttachedDbs();
    }

    @Override
    public boolean isDatabaseIntegrityOk() {
        return db.isDatabaseIntegrityOk();
    }

    @Override
    public void close() throws IOException {
        db.close();
    }

    public void insertWithOnConflict(String tableName, String nullColumnHack, ContentValues values, int conflictIgnore) {
        db.insertWithOnConflict(tableName, nullColumnHack, values, conflictIgnore);
    }

    public Cursor rawQuery(String sql, Object[] selectionArgs) {
        return db.rawQuery(sql, selectionArgs);
    }

    public interface QueryLogger {
        void onQuery(String query);
    }
}

File: src/main/java/com/gazman/db/DbProxy.kt
package com.gazman.db

import android.content.ContentValues
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteTransactionListener
import android.os.Build
import android.os.CancellationSignal
import androidx.annotation.RequiresApi
import androidx.sqlite.db.SupportSQLiteDatabase
import androidx.sqlite.db.SupportSQLiteQuery
import androidx.sqlite.db.SupportSQLiteStatement
import io.requery.android.database.sqlite.SQLiteDatabase
import java.io.IOException
import java.util.*

class DbProxy(
    private val db: SupportSQLiteDatabase
) : SupportSQLiteDatabase {

    override val attachedDbs = db.attachedDbs
    override val isDatabaseIntegrityOk = db.isDatabaseIntegrityOk
    override val isDbLockedByCurrentThread = db.isDbLockedByCurrentThread
    override val isOpen = db.isOpen
    override val isReadOnly = db.isReadOnly
    override val isWriteAheadLoggingEnabled = db.isWriteAheadLoggingEnabled
    override val maximumSize = db.maximumSize
    override var pageSize = db.pageSize
    override val path = db.path
    override var version = db.version


    override fun compileStatement(sql: String): SupportSQLiteStatement {
        return db.compileStatement(sql)
    }

    override fun delete(table: String, whereClause: String?, whereArgs: Array<out Any?>?) =
        db.delete(table, whereClause, whereArgs)

    override fun beginTransaction() {
        db.beginTransaction()
    }

    override fun beginTransactionNonExclusive() {
        db.beginTransactionNonExclusive()
    }

    override fun beginTransactionWithListener(transactionListener: SQLiteTransactionListener) {
        db.beginTransactionWithListener(transactionListener)
    }

    override fun beginTransactionWithListenerNonExclusive(transactionListener: SQLiteTransactionListener) {
        db.beginTransactionWithListenerNonExclusive(transactionListener)
    }

    override fun endTransaction() {
        db.endTransaction()
    }

    override fun setTransactionSuccessful() {
        db.setTransactionSuccessful()
    }

    override fun update(
        table: String,
        conflictAlgorithm: Int,
        values: ContentValues,
        whereClause: String?,
        whereArgs: Array<out Any?>?
    ) = db.update(table, conflictAlgorithm, values, whereClause, whereArgs)

    override fun inTransaction(): Boolean {
        return db.inTransaction()
    }

    override fun yieldIfContendedSafely(): Boolean {
        return db.yieldIfContendedSafely()
    }

    override fun yieldIfContendedSafely(sleepAfterYieldDelayMillis: Long): Boolean {
        return db.yieldIfContendedSafely(sleepAfterYieldDelayMillis)
    }

    override fun setMaximumSize(numBytes: Long): Long {
        return db.setMaximumSize(numBytes)
    }

    override fun query(query: String): Cursor {
        log(query)
        return db.query(query)
    }

    override fun query(query: String, bindArgs: Array<out Any?>): Cursor {
        log(query)
        return db.query(query, bindArgs)
    }

    override fun query(query: SupportSQLiteQuery): Cursor {
        return db.query(query)
    }

    override fun query(query: SupportSQLiteQuery, cancellationSignal: CancellationSignal?) =
        db.query(query, cancellationSignal)

    @Throws(SQLException::class)
    override fun insert(table: String, conflictAlgorithm: Int, values: ContentValues): Long {
        return db.insert(table, conflictAlgorithm, values)
    }

    fun insert(table: String, values: ContentValues): Long {
        return db.insert(table, SQLiteDatabase.CONFLICT_NONE, values)
    }

    fun delete(table: String) = db.delete(table, null, null)
    fun delete(table: String, whereClause: String?) = db.delete(table, whereClause, null)

    fun update(
        table: String,
        values: ContentValues,
        whereClause: String?,
        whereArgs: Array<String?>?
    ): Int {
        return db.update(table, SQLiteDatabase.CONFLICT_NONE, values, whereClause, whereArgs)
    }

    @Throws(SQLException::class)
    override fun execSQL(sql: String) {
        log(sql)
        db.execSQL(sql)
    }

    override fun execSQL(sql: String, bindArgs: Array<out Any?>) {
        db.execSQL(sql, bindArgs)
    }

    private fun log(query: String) {
        queryLogger?.onQuery(query)
    }

    override fun needUpgrade(newVersion: Int): Boolean {
        return db.needUpgrade(newVersion)
    }

    override fun setLocale(locale: Locale) {
        db.setLocale(locale)
    }

    override fun setMaxSqlCacheSize(cacheSize: Int) {
        db.setMaxSqlCacheSize(cacheSize)
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    override fun setForeignKeyConstraintsEnabled(enabled: Boolean) {
        db.setForeignKeyConstraintsEnabled(enabled)
    }

    override fun enableWriteAheadLogging(): Boolean {
        return db.enableWriteAheadLogging()
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    override fun disableWriteAheadLogging() {
        db.disableWriteAheadLogging()
    }

    @Throws(IOException::class)
    override fun close() {
        db.close()
    }

    fun insertWithOnConflict(
        tableName: String,
        values: ContentValues,
        conflictIgnore: Int
    ) {
        db.insert(tableName, conflictIgnore, values)
    }

    fun rawQuery(sql: String, selectionArgs: Array<Any?>? = null): Cursor {
        log(sql)
        return if (selectionArgs != null) {
            db.query(sql, selectionArgs)
        } else {
            db.query(sql)
        }
    }

    fun interface QueryLogger {
        fun onQuery(query: String?)
    }

    companion object {
        const val CONFLICT_IGNORE = SQLiteDatabase.CONFLICT_IGNORE
        const val CONFLICT_REPLACE = SQLiteDatabase.CONFLICT_REPLACE
        var queryLogger: QueryLogger? = null
    }
}